data_pipeline/generic_web_api/generic_web_api_data_etl.py

```diff
@@ -0,0 +1,451 @@
+    ):
+        from_timestamp = current_start_timestamp
+    elif (
+            web_config.item_timestamp_key_hierarchy_from_item_root and
+            items_count
+    ):
+        from_timestamp = latest_record_timestamp
+
+    return from_timestamp
+
+
+def get_next_cursor_from_data(data, web_config: WebApiConfig):
+    next_cursor = None
+    if web_config.url_manager.next_page_cursor:
+        next_cursor = get_dict_values_from_hierarchy_as_list(
+            data,
+            web_config.next_page_cursor_key_hierarchy_from_response_root
+        )
+    return next_cursor
+
+
+def get_items_list(data, web_config):
+    item_list = data
+    if isinstance(data, dict):
+        item_list = get_dict_values_from_hierarchy_as_list(
+            data,
+            web_config.items_key_hierarchy_from_response_root
+        )
+    return item_list
+
+
+def upload_latest_timestamp_as_pipeline_state(
+        data_config,
+        latest_record_timestamp: datetime
+):
+    if (
+            data_config.state_file_object_name and
+            data_config.state_file_bucket_name
+    ):
+        latest_record_date = get_timestamp_as_string(latest_record_timestamp)
+        state_file_name_key = data_config.state_file_object_name
+        state_file_bucket = data_config.state_file_bucket_name
+        upload_s3_object(
+            bucket=state_file_bucket,
+            object_key=state_file_name_key,
+            data_object=latest_record_date,
+        )
+
+
+def create_or_extend_table_schema(
+        data_config: WebApiConfig,
+        full_temp_file_location,
+):
+    schema = generate_schema_from_file(
+        full_temp_file_location
+    )
+
+    if does_bigquery_table_exist(
+            data_config.gcp_project,
+            data_config.dataset_name,
+            data_config.table_name,
+    ):
+        extend_table_schema_with_nested_schema(
+            data_config.gcp_project,
+            data_config.dataset_name,
+            data_config.table_name,
+            schema
+        )
+    else:
+        create_table(
+            data_config.gcp_project,
+            data_config.dataset_name,
+            data_config.table_name,
+            schema
+        )
+
+
+def get_bq_schema(data_config: WebApiConfig,):
+    if (
+            data_config.schema_file_object_name and
+            data_config.schema_file_s3_bucket
+    ):
+        bq_schema = download_s3_json_object(
+            data_config.schema_file_s3_bucket,
+            data_config.schema_file_object_name
+        )
+    else:
+        bq_schema = None
+    return bq_schema
+
+
+def process_downloaded_data(
+        record_list: list,
+        data_config: WebApiConfig,
+        data_etl_timestamp,
+        file_location,
+        prev_page_latest_timestamp: datetime = None
+):
+    provenance = {
+        data_config.import_timestamp_field_name:
+            data_etl_timestamp
+    }
+    bq_schema = get_bq_schema(data_config)
+    n_record_list = process_record_in_list(
```

Comment:
minor: might be worth if we could replace the `n` (wherever it appears) with something more meaningful.